### ISSUE TITLE
MPI bootstrap issue on Perlmutter

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#5375]
A user on Perlmutter is trying to use a spack installed version of flux 

```
flux --version
commands:            0.49.0
libflux-core:        0.49.0
broker:          0.49.0
FLUX_URI:        local:///tmp/flux-15AFKQ/local-0
build-options:        +hwloc==2.8.0+zmq==4.3.4
```

to launch an mpich executable and it is apparently launching in singleton mode (evidence is that the output is printed 16 times instead of once).

@nkeilbart - if you could paste the output of `flux run -o verbose=2 -n16 EXECUTABLE` here then that might be a good place to start debugging. Thanks!

**Comment by grondo** [comment]
Also, if you can, please include the version of mpich linked in your application. Thanks!

**Comment by nkeilbart** [comment]
0.795s: flux-shell[0]: DEBUG: Loading /global/u2/k/keilbart/git/Perlmutter/spack/opt/spack/linux-sles15-zen3/gcc-11.2.0/flux-core-0.49.0-zvwbgtouotpvr4lptn2byfqanvw2bd3x/etc/flux/shell/initrc.lua
0.801s: flux-shell[0]: TRACE: Sucessfully loaded flux.shell module
0.801s: flux-shell[0]: TRACE: trying to load /global/u2/k/keilbart/git/Perlmutter/spack/opt/spack/linux-sles15-zen3/gcc-11.2.0/flux-core-0.49.0-zvwbgtouotpvr4lptn2byfqanvw2bd3x/etc/flux/shell/initrc.lua
0.806s: flux-shell[0]: TRACE: trying to load /global/u2/k/keilbart/git/Perlmutter/spack/opt/spack/linux-sles15-zen3/gcc-11.2.0/flux-core-0.49.0-zvwbgtouotpvr4lptn2byfqanvw2bd3x/etc/flux/shell/lua.d/intel_mpi.lua
0.809s: flux-shell[0]: TRACE: trying to load /global/u2/k/keilbart/git/Perlmutter/spack/opt/spack/linux-sles15-zen3/gcc-11.2.0/flux-core-0.49.0-zvwbgtouotpvr4lptn2byfqanvw2bd3x/etc/flux/shell/lua.d/openmpi.lua
0.811s: flux-shell[0]: DEBUG: pmi-simple: simple wire protocol is enabled
0.813s: flux-shell[0]: DEBUG: output: batch timeout = 0.500s
0.815s: flux-shell[0]: DEBUG: 0: task_count=16 slot_count=16 cores_per_slot=1 slots_per_node=-1
0.815s: flux-shell[0]: DEBUG: 0: tasks [0-15] on cores 48-63
MPI initialized with 1 MPI processes
MPI initialized with 1 MPI processes
MPI initialized with 1 MPI processes
MPI initialized with 1 MPI processes
MPI initialized with 1 MPI processes
MPI initialized with thread support level 0
MPI initialized with thread support level 0
MPI initialized with thread support level 0
MPI initialized with thread support level 0
MPI initialized with thread support level 0
MPI initialized with 1 MPI processes
MPI initialized with thread support level 0
MPI initialized with 1 MPI processes
MPI initialized with thread support level 0
MPI initialized with 1 MPI processes
MPI initialized with thread support level 0
MPI initialized with 1 MPI processes
MPI initialized with thread support level 0
AMReX (23.08-21-gfdce60c55e7e) initialized
AMReX (23.08-21-gfdce60c55e7e) initialized
AMReX (23.08-21-gfdce60c55e7e) initialized
AMReX (23.08-21-gfdce60c55e7e) initialized
AMReX (23.08-21-gfdce60c55e7e) initialized
AMReX (23.08-21-gfdce60c55e7e) initialized
AMReX (23.08-21-gfdce60c55e7e) initialized
AMReX (23.08-21-gfdce60c55e7e) initialized
AMReX (23.08-21-gfdce60c55e7e) initialized
MPI initialized with 1 MPI processes
MPI initialized with thread support level 0
AMReX (23.08-21-gfdce60c55e7e) initialized
MPI initialized with 1 MPI processes
MPI initialized with thread support level 0
AMReX (23.08-21-gfdce60c55e7e) initialized
MPI initialized with 1 MPI processes
MPI initialized with thread support level 0
AMReX (23.08-21-gfdce60c55e7e) initialized
MPI initialized with 1 MPI processes
MPI initialized with thread support level 0
AMReX (23.08-21-gfdce60c55e7e) initialized
MPI initialized with 1 MPI processes
MPI initialized with thread support level 0
AMReX (23.08-21-gfdce60c55e7e) initialized
MPI initialized with 1 MPI processes
MPI initialized with thread support level 0
MPI initialized with 1 MPI processes
MPI initialized with thread support level 0
AMReX (23.08-21-gfdce60c55e7e) initialized
AMReX (23.08-21-gfdce60c55e7e) initialized
Total number of agents: 300072297
Splitting up population into interior and border
Total number of agents: 300072297
Splitting up population into interior and border
Total number of agents: 300072297
Splitting up population into interior and border
Total number of agents: 300072297
Splitting up population into interior and border
Total number of agents: 300072297
Splitting up population into interior and border
Total number of agents: 300072297
Splitting up population into interior and border
Total number of agents: 300072297
Splitting up population into interior and border
Total number of agents: 300072297
Splitting up population into interior and border
Total number of agents: 300072297
Splitting up population into interior and border
Total number of agents: 300072297
Splitting up population into interior and border
Total number of agents: 300072297
Total number of agents: 300072297
Splitting up population into interior and border
Splitting up population into interior and border
Total number of agents: 300072297
Splitting up population into interior and border
Total number of agents: 300072297
Splitting up population into interior and border
Total number of agents: 300072297
Splitting up population into interior and border
Total number of agents: 300072297
Splitting up population into interior and border
Population within 200 cells of border is 110026244
Population within 200 cells of border is 110026244
Population within 200 cells of border is 110026244
Population within 200 cells of border is 110026244
Population within 200 cells of border is 110026244
Population within 200 cells of border is 110026244
Population within 200 cells of border is 110026244
Population within 200 cells of border is 110026244
Population within 200 cells of border is 110026244
Population within 200 cells of border is 110026244
Population within 200 cells of border is 110026244
Population within 200 cells of border is 110026244
Population within 200 cells of border is 110026244
Population within 200 cells of border is 110026244
Population within 200 cells of border is 110026244
Population within 200 cells of border is 110026244
36.319s: flux-shell[0]: TRACE: pmi-simple: 4: C: pmi EOF
36.320s: flux-shell[0]: DEBUG: task 4 complete status=137
Initial Redistribute... 42.470s: flux-shell[0]: TRACE: pmi-simple: 9: C: pmi EOF
42.470s: flux-shell[0]: DEBUG: task 9 complete status=137
Initial Redistribute... 49.565s: flux-shell[0]: TRACE: pmi-simple: 13: C: pmi EOF
49.565s: flux-shell[0]: DEBUG: task 13 complete status=137
Initial Redistribute... 57.513s: flux-shell[0]: TRACE: pmi-simple: 0: C: pmi EOF
57.513s: flux-shell[0]: DEBUG: task 0 complete status=137
Initial Redistribute... 66.334s: flux-shell[0]: FATAL: doom: rank 4 exited and exit-timeout=30s has expired
66.344s: job.exception type=exec severity=0 rank 4 exited and exit-timeout=30s has expired
flux-job: task(s) exited with exit code 137

**Comment by nkeilbart** [comment]
Looks like it should be this version: cray-mpich/8.1.25

**Comment by garlick** [comment]
If you can find the directory where flux's `libpmi2.so` is installed, prepending that directory name to the LD_LIBRARY_PATH may help.  I _think_ this should work:
```
$ dirname $(flux getattr conf.pmi_library_path)
/usr/lib/aarch64-linux-gnu/flux
```
But of course the path will be different for your spack installation.  Then something like
```
$ LD_LIBRARY_PATH=/usr/lib/aarch64-linux-gnu/flux:$LD_LIBRARY_PATH flux run -o verbose=2 -n16 EXECUTABLE
```

**Comment by garlick** [comment]
Mentioning @jameshcorbett since I think he's had some experience running flux on perlmutter (though probably a while back!)

**Comment by nkeilbart** [comment]
Thanks that seems to solve that issue. I can simply add that to my start_flux.sh script for now. Is there some reason that it's not being added to LD_LIBRARY_PATH or some other thing I might have done incorrectly when installing?

Also, should we open a different issue for flux not properly detecting the GPUs when being launched with srun?

**Comment by nkeilbart** [comment]
It seems in my tired state from a trip I had to make I forgot to ask for the gpus to also be assigned to the gpu allocation on Perlmutter. I am able to get the GPUs once I set that. 

The LD_LIBRARY_PATH has fixed things for me currently. Is there a possible permanent solution so that I won't need to add that every time?

**Comment by grondo** [comment]
You can export the environment variable once in the environment of the shell after `flux start`. (I'm not sure what would happen if you did it before, I think perhaps `srun` would not be able to launch Flux). This can be convenient enough if you're 

Another solution would be to adjust your `start_flux.sh` to use this for the `flux start` line:
```sh
 flux start bash -c 'export LD_LIBRARY_PATH="$(dirname $(flux getattr conf.pmi_library_path)):$LD_LIBRARY_PATH" && exec $SHELL'
```

This will just automatically export `LD_LIBRARY_PATH` then invoke the current shell (same as what would happen with `flux start` alone).  If you are already calling `flux start` with arguments, use this instead:
```sh
 flux start bash -c 'export LD_LIBRARY_PATH="$(dirname $(flux getattr conf.pmi_library_path)):$LD_LIBRARY_PATH" && exec "$@"' ARGS
```

**Comment by grondo** [comment]
Verified that this is fixed with the `LD_LIBRARY_PATH` workaround.

**Comment by garlick** [comment]
Note that #5715 should have made the LD_LIBRARY_PATH hack unnecessary.  This change was released with flux-core-0.59.0.
